apps/api-extractor-model/src/mixins/ApiTypeParameterListMixin.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.s
+
+import { ApiItem, IApiItemJson, IApiItemConstructor, IApiItemOptions } from '../items/ApiItem';
+import { IExcerptTokenRange } from './Excerpt';
+import { TypeParameter } from '../model/TypeParameter';
+import { InternalError } from '@microsoft/node-core-library';
+import { ApiDeclaredItem } from '../items/ApiDeclaredItem';
+
+/**
+ * Represents parameter information that is part of {@link IApiTypeParameterListMixinOptions}
+ * @public
+ */
+export interface IApiTypeParameterOptions {
+  typeParameterName: string;
+  constraintTokenRange?: IExcerptTokenRange;
+  defaultTypeTokenRange?: IExcerptTokenRange;
```

Comment:
If you compare with `IApiParameterOptions`, you'll see that its `parameterTypeTokenRange` is not optional, even though it's possible for the type to be missing as with `x` in this example:

```ts
declare function f(x, y: number): void;
```

Similarly the `IApiParameterListMixinOptions.parameters` list is also not optional, even though a function may have 0 parameters.

Thus we redundantly specify the array even if it has zero items, and redundantly specify an empty `IExcerptTokenRange` even if there is no parameter type, and even redundantly write an empty token range into the .api.json file.  The rationale for this design was that the consumer code such as **ApiModelGenerator.ts** tends to be fairly complicated and confusing, so it's easy to accidentally forget to pass along one of the inputs if it's not mandatory.  Better to make the person pass an empty placeholder, so they can't omit the thing without stopping to ask themselves /why/ they're making an empty thing.  (Did they forget to capture part of the signature?)

The part about _"redundantly write an empty token range into the .api.json file"_ is debatable.  Back with API Extractor 6, consumers were writing their own code to load and parse the JSON; for that scenario, always writing every field made life much easier since the JSON schema could be easily inferred from the file itself.  But with the release of API Extractor 7, the **api-extractor-model** library now provides a standard deserializer, so nobody should be writing their own .api.json parser.  Thus we could perhaps make the serialized file significantly smaller by omitting empty fields such as `IApiTypeParameterListMixinJson.typeParameters.constraintTokenRange` or even `IApiTypeParameterListMixinJson.typeParameters`.  But my leaning would still be to approach such optimizations conservatively, because they add more schema permutations which may increase our test surface.

I'm open-minded about omitting unused fields from the .api.json file. But unless there's a strong counterargument, I'd feel safer keeping `IApiTypeParameterListMixinOptions.typeParameters`, `IApiTypeParameterOptions.constraintTokenRange`, etc as non-optional.

Reply:
> Thus we redundantly specify the array even if it has zero items, and redundantly specify an empty IExcerptTokenRange even if there is no parameter type, and even redundantly write an empty token range into the .api.json file.

However, `IApiClassOptions` specifies `extendsTokenRange` as allowing `undefined`:

```ts
export interface IApiClassOptions extends
  IApiItemContainerMixinOptions,
  IApiNameMixinOptions,
  IApiReleaseTagMixinOptions,
  IApiDeclaredItemOptions,
  IApiTypeParameterListMixinOptions {

  extendsTokenRange: IExcerptTokenRange | undefined;
  implementsTokenRanges: IExcerptTokenRange[];
}
```

Perhaps this is an oversight?

Reply:
Well it is possibly undefined but not optional. I suppose that accomplishes the purpose of making sure people don't forget to pass the parameter. Maybe we should apply that approach to `Parameter` and `TypeParameter` too? I haven't worked with this code in awhile so I don't remember the exact usage details. Just that I was really paranoid about forgetting to pass along an input. :-)

Reply:
I think it's fine to leave them as defined. I added an `isEmpty` property to `Excerpt` to make it easier to test whether or not the except contains no tokens to make this check easier.

Reply:
> However, `IApiClassOptions` specifies `extendsTokenRange` as allowing `undefined`:

I spent a little more time investigating this.  From an API design perspective, they're the same kind of thing and the usage should be consistent.  However, there's some plumbing issues that got us to the current design:

It started with the idea that `Excerpt` is like an array of tokens, and arrays can have zero elements.  If we allow `Except` properties to be `undefined`, then is there a semantic difference between that versus an empty array?  Probably not.  And if not, why should we have two states that mean the same thing?  Thus I chose to introduce `Excerpt.isEmpty()` and make a convention that `Excerpt` properties are never `undefined`, which means any code that loops over the array can avoid having to check for `undefined`.  

This is all in the world of `Parameter`, not `IParameterOptions`, but it seemed reasonable to apply the same convention with `IExcerptTokenRange`, passing an empty range instead of `undefined`.  In fact it "helped" by enabling [ExcerptBuilder.build()](https://github.com/microsoft/web-build-tools/blob/d85ea6f5d9e7ba5ba8e136dc1aabb4e5fcdeadd1/apps/api-extractor/src/generators/ExcerptBuilder.ts#L72) to mutate the empty `IExcerptTokenRange` that it receives:

https://github.com/microsoft/web-build-tools/blob/d85ea6f5d9e7ba5ba8e136dc1aabb4e5fcdeadd1/apps/api-extractor/src/generators/ExcerptBuilder.ts#L144-L152

This somewhat weird design enables `_captureParameters()` to construct the `IApiParameterOptions` before the `ExcerptBuilder.build()` has actually assigned their values, avoiding an awkward postprocessing pass:

https://github.com/microsoft/web-build-tools/blob/d85ea6f5d9e7ba5ba8e136dc1aabb4e5fcdeadd1/apps/api-extractor/src/generators/ApiModelGenerator.ts#L704-L717

In this context, an empty range means the `ExcerptBuilder` didn't find a match.  In retrospect this design feels not great. 😋  If someone can think of a better way to enable `_captureParameters()` to make the `IApiParameterOptions` before the `ExcerptBuilder.build()` goes scanning for tokens, I'd be interested in that.

The last piece of the puzzle is `IApiClassOptions.extendsTokenRange` which you mentioned.  In this case it doesn't produce an `Excerpt` but rather a `HeritageType`:

https://github.com/microsoft/web-build-tools/blob/d85ea6f5d9e7ba5ba8e136dc1aabb4e5fcdeadd1/apps/api-extractor-model/src/model/ApiClass.ts#L49-L52

Following our convention, the `HeritageType.excerpt` cannot be `undefined`.  When `IApiClassOptions.extendsTokenRange` is `undefined` this really means that `ApiClass.extendsType` will be `undefined`.  And it's worth pointing out that `extends` changes the semantics of the code; it's not just a type annotation.  So perhaps it's wise to shy away from using empty ranges here.

The overall design choices here seem like they can be improved, and maybe we can eliminate the passing of empty ranges entirely.  I'm interested to do that, but it's out of scope for your PR here, which is now following the existing precedent.  These improvements should be a separate PR.